(10) Zahoda

RinRowRow opened the pull request.

kolmogorov-aa commented on solutions/src/main/java/ru/sberbank/school/task10/FixedThreadPool.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Override
+    public void start() {
```
Два раза стартануть тредпул. нехорошо будет

RinRowRow replied:
Так у меня же с threads.size() цикл начинается.

RinRowRow replied:
Впрочем. Я так понял, что в целом нельзя позволять 2 раза запускать. Так?

RinRowRow replied:
Ну. Без остановки.

kolmogorov-aa commented on solutions/src/main/java/ru/sberbank/school/task10/FixedThreadPool.java:
```diff
@@ -0,0 +1,86 @@
+    @Override
+    public void execute(@NonNull Runnable runnable) {
+        synchronized (taskQueue) {
+            taskQueue.add(runnable);
```
ну неее... Оно вообще работает?

RinRowRow replied:
Ааааа. Я правильно понял, что execute можно только после старта вызывать? Я просто думал, что так можно как бы в очередь складывать. Ну тогда понял проблему. Сейчас решу).

kolmogorov-aa commented on solutions/src/main/java/ru/sberbank/school/task10/FixedThreadPool.java:
```diff
@@ -0,0 +1,86 @@
+
+        @Override
+        public void run() {
+            while (!Thread.interrupted()) {
```
этот цикл непрерывно работает в каждом треде, так получается?
Busy waiting не хорошо

kolmogorov-aa commented on solutions/src/main/java/ru/sberbank/school/task10/ScalableThreadPool.java:
```diff
@@ -0,0 +1,111 @@
+                    if (taskQueue.isEmpty() && workingThreads > min) {
+                        idsManager.setFreeId(id);
+                        workingThreads--;
+                        threads.remove(this);
```
Похоже, что мы имеем дело с самоубийством =)

kolmogorov-aa reviewed (changes requested):
Не могу замержить в таком состоянии

RinRowRow committed:
id 10

Добавил сборщик тредов)

Подправил сборщик потоков

Добавил mainLock и подправил тесты.

Добавил тест и отредактировал пулы.

Remove empty catch block.

kolmogorov-aa commented on solutions/src/main/java/ru/sberbank/school/task10/ScalableThreadPool.java:
```diff
@@ -0,0 +1,180 @@
+
+    @Override
+    public void execute(@NonNull Runnable runnable) {
+        synchronized (mainLock) {
```
зачем нужен main lock?

RinRowRow replied:
Точно. Там достаточно было добавить synchronized в методы. Не подумал)

kolmogorov-aa commented on solutions/src/main/java/ru/sberbank/school/task10/ScalableThreadPool.java:
```diff
@@ -0,0 +1,180 @@
+                    try {
+                        freeThreadLock.wait();
+                        idsManager.addFreeId(freeThread.getThreadId());
+                        synchronized (threads) {
```
Беда у вас с блокировками их, как мне кажется, очень много. Много блокировок, много мониторов.

по крайней мере мне так показалось, сложно анализировать такой код

RinRowRow replied:
У меня была идея с демон потоком, который раз в опр время чистит неиспользуемые потоки, но тогда бы пришлось в нем или проходить по всем тредам и проверять на isAlive, или создать отдельный список "освобожденных" потоков, но тогда непонятно, как их в этот список добавлять(. Наверное, мне нужно подумать совсем в другом ключе.

kolmogorov-aa commented on solutions/src/main/java/ru/sberbank/school/task10/ScalableThreadPool.java:
```diff
@@ -0,0 +1,180 @@
+                synchronized (freeThreadLock) {
+                    try {
+                        freeThreadLock.wait();
+                        idsManager.addFreeId(freeThread.getThreadId());
```
это на грани фантастики... 
очень костыльно выглядит. 

это не страшно на текущем этапе обучения, но не стоит такие техники практиковать.
